src/main/java/net/minecraftforge/client/event/RenderLevelStageEvent.java

```diff
@@ -0,0 +1,145 @@
+         *     If null, needs to be called manually by whoever implements it.
+         * 
+         */
+        public Stage register(ResourceLocation name, @Nullable RenderType renderType)
+        {
+            return Stage.register(name, renderType);
+        }
+    }
+
+    /**
+     * A time during level rendering for you to render custom things into the world.
+     * @see RegisterStageEvent
+     */
+    public static class Stage
+    {
+        private static final Map<RenderType, Stage> RENDER_TYPE_STAGES = new HashMap<>();
+
+        /**
+         * Use this to render custom objects into the skybox.
+         * Called regardless of if they sky actually renders or not.
+         */
+        public static final Stage AFTER_SKY = register("after_sky", null);
+        /**
+         * Use this to render custom block-like geometry into the world.
+         */
+        public static final Stage AFTER_SOLID_BLOCKS = register("after_solid_blocks", RenderType.solid());
+        /**
+         * Use this to render custom block-like geometry into the world.
+         */
+        public static final Stage AFTER_CUTOUT_MIPPED_BLOCKS_BLOCKS = register("after_cutout_mipped_blocks", RenderType.cutoutMipped());
+        /**
+         * Use this to render custom block-like geometry into the world.
+         */
+        public static final Stage AFTER_CUTOUT_BLOCKS = register("after_cutout_blocks", RenderType.cutout());
+        /**
+         * Use this to render custom block-like geometry into the world.
+         * Called within a fabulous graphics target. 
+         * Happens after block entities render. 
+         */
+        public static final Stage AFTER_TRANSLUCENT_BLOCKS = register("after_translucent_blocks", RenderType.translucent());
+        /**
+         * Use this to render custom block-like geometry into the world.
+         */
+        public static final Stage AFTER_TRIPWIRE_BLOCKS = register("after_tripwire_blocks", RenderType.tripwire());
+        /**
+         * Use this to render custom effects into the world, such as custom entity-like objects or special rendering effects.
+         * Called within a fabulous graphics target.
+         * Happens after entities render.
+         */
+        public static final Stage AFTER_PARTICLES = register("after_particles", null);
+        /**
+         * Use this to render custom weather effects into the world.
+         * Called within a fabulous graphics target.
+         */
+        public static final Stage AFTER_WEATHER = register("after_weather", null);
+
+        private final String name;
+
+        private Stage(String name)
+        {
+            this.name = name;
+        }
+
+        private static Stage register(ResourceLocation name, @Nullable RenderType renderType)
+        {
+            Stage stage = new Stage(name.toString());
+            if (renderType != null && RENDER_TYPE_STAGES.put(renderType, stage) != null)
```

Comment:
Use [`Map#putIfAbsent`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Map.html#putIfAbsent(K,V)) instead of [`Map#put`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/Map.html#put(K,V)), as `putIfAbsent` only modifies the map if they key had no value previously, while `put` modifies the map regardless of whether or not the key had a value previously.